Saved var refactor [IDB IGNORE] [MDB IGNORE]

PsyCommando opened the pull request:
## Description of changes
Moved all saved vars handling to code since the saved_var.json was severely out of date and unmaintainable. 

Now use a system with decl to store saved vars for each individual saved types. Which allows for compile time validation and var checking.

MarinaGryphon committed:
Fixes timer GC

MarinaGryphon committed:
Reorder atom/movable/Destroy to fix GC

MarinaGryphon committed:
Expand GC fail reference searching

MarinaGryphon committed:
Fix mind and mob GC

MarinaGryphon committed:
Fix autosay being garbled

MarinaGryphon committed:
Fix grammar in unpowered computer message

MarinaGryphon committed:
Fix several more hanging refs

MarinaGryphon committed:
Clear some hanging references on atoms/prefs/mobs

MarinaGryphon committed:
Fix hanging refs on vines

MarinaGryphon committed:
Update SSgarbage ref finding

MarinaGryphon committed:
Fix circular reference in thrownthing

NataKilar committed:
Network access rework

quardbreak committed:
remaps example map

quardbreak committed:
repaths stairs in map files

quardbreak committed:
adds empty png files for test site minimap

NataKilar committed:
Network access adjustments

NataKilar committed:
Further network adjustments

NataKilar committed:
Persistence access implementation

noelle-lavenza committed:
Fix tail icons

PsyCommando committed:
When spawning, the greeting messages check if you have a headset

Greeting message now check if your starting loadout actually gives you a headset before advising you on how to talk on your headset..

MistakeNot4892 committed:
Merge pull request #2317 from PsyCommando/greeting-msg-check-headset-first

When spawning, the greeting messages check if you have a headset

MistakeNot4892 committed:
Merge pull request #2311 from quardbreak/sstimer-timer-max

timer fixes from ss220

NebulaSS13Bot committed:
Automatic changelog generation for PR #2317 [ci skip]

MistakeNot4892 committed:
Merge pull request #2269 from quardbreak/dmdoc

dmdoc: adds workflow for dev branch

MistakeNot4892 committed:
Merge pull request #2267 from quardbreak/keybinds-fixes

keybind tweaks

quardbreak committed:
review changes

NebulaSS13Bot committed:
Automatic changelog generation [ci skip]

afterthought2 committed:
Merge pull request #2271 from quardbreak/bunchofjunk

New year's eve fixes

quardbreak committed:
re-examplenate example map

MistakeNot4892 committed:
Prevented skills taking over generic codex strings.

MistakeNot4892 committed:
Added codex ID overlap check unit test.

MistakeNot4892 committed:
Cleaning up some recipes/paths/overlapping codex entry names.

MistakeNot4892 committed:
Pruning some extraneous codex strings.

MistakeNot4892 committed:
Further work on cleaning up overlapping codex pages.

MistakeNot4892 committed:
Adding a disambiguator to codex definitions.

MistakeNot4892 committed:
Tweaking recipes.

MistakeNot4892 committed:
Disambiguating substances.

MistakeNot4892 committed:
Tweaking metals.

MistakeNot4892 committed:
Converts spell codex entries to an autogen category.

MistakeNot4892 committed:
More substance tweaks.

MistakeNot4892 committed:
Added codex string uniqueness checking.

MistakeNot4892 committed:
Even more substance codex name tweaks.

MistakeNot4892 committed:
Ticks phenomena category.

MistakeNot4892 committed:
Reworking/cleaning up codex entry creation.

MistakeNot4892 committed:
Cleaning up more associated strings.

MistakeNot4892 committed:
Still more disambiguation.

MistakeNot4892 committed:
Cleaning up categories.

MistakeNot4892 committed:
Trying to chase down odd duplications in associated_strings list.

MistakeNot4892 committed:
Codex categories no longer initialize twice.

MistakeNot4892 committed:
Refining codex_name entries.

quardbreak committed:
Fixes reference finder on globals

PsyCommando committed:
Rejuvenating the heart should now fix heart stopping

PsyCommando committed:
Added check for prosthetic heart

Prosthetic hearts don't have a pulse.

NataKilar committed:
Computer login automatically reads from ID

NataKilar committed:
Restricts network group naming

NataKilar committed:
Reports and network lock adjustments

PsyCommando committed:
Tweaked cable laying a bit

* All turfs now use try_build_cable for attempting to build a cable.
* try_build_cable warns the user if something is off instead of can_build_cable
* can_build_cable now actually only checks if we can build a cable or not, and doesn't actually prints text to the user anymore. So it won't spam the users when just doing a check.
* can_build_cable is now defined in /turf since cable can be installed on several branching types of turfs. Same with try_build_cable.

PsyCommando committed:
Added feedback when trying to use a cable coil on a zcable

Now instead of doing nothing, the cable gets connected the way zcable expect, and a message is sent to the user to let them know that yes, even if it doesn't look like it, the cable are indeed connected.

PsyCommando committed:
Corrected old path typo

Seems pretty clear that this proc wasn't meant to be overridden twice on /turf/simulated.

PsyCommando committed:
General messages cleanup

* Used macros for spans.
* Used defined stack vars to refer to the stack's contents instead of just using different words each times.
* Replaced large unreadable power numbers with power constants instead.

MistakeNot4892 committed:
Fixes misplaced square brackets.

MistakeNot4892 committed:
Fixes broken offset define.

quardbreak committed:
destroy: ..() to return ..()

afterthought2 committed:
Merge pull request #2329 from quardbreak/find_references_global

Fixes reference finder on globals

MistakeNot4892 committed:
Adds multi-init checking to decls.

MistakeNot4892 committed:
Minor inflatables refactor, makes inflatables more available/craftable.

afterthought2 committed:
Merge pull request #2338 from quardbreak/destroy-return

destroy: ..() to return ..()

quardbreak committed:
Removes number 13 from game version string

quardbreak committed:
Fixes incorrect game id generation

MistakeNot4892 committed:
Standardizing robotize, adding capacity to check against specific species/body (for downstream).

MistakeNot4892 committed:
Blast doors now autoset direction.

MistakeNot4892 committed:
Condensing grab logic in /mob/living/Move().

MistakeNot4892 committed:
Adjusts ladder handling to allow grabs.

MistakeNot4892 committed:
Gating off a division by zero check. Unsure of source, showing up in unit testing.

NataKilar committed:
Fixes "or" text on network locks

MistakeNot4892 committed:
Adds missing blood type to species def.

MistakeNot4892 committed:
Prevents parrots calming down eternally.

afterthought2 committed:
Merge pull request #2344 from MistakeNot4892/gas

Gating off a division by zero check in gas heat capacity.

afterthought2 committed:
Merge pull request #2340 from MistakeNot4892/armfix

Standardizing robotize, adding capacity to check against specific species/bodytype.

PsyCommando committed:
Added godmode checks to fall damage + dislocating + shock (#2336)

* Added godmode checks to fall damage + dislocating

* Added missing parenthesis

NebulaSS13Bot committed:
Automatic changelog generation for PR #2336 [ci skip]

MistakeNot4892 committed:
Blacklisted abstract material types from exoplanet gas selection.

afterthought2 committed:
Merge pull request #2341 from MistakeNot4892/doors

Blast doors now autoset direction from neighbors.

MistakeNot4892 committed:
Removing broken vendor icon.

MistakeNot4892 committed:
Adds shields to industrial fab, allows buckler crafting.

MistakeNot4892 committed:
Resets dislocation in rejuvenate.

MistakeNot4892 committed:
Shuffling around the hardness values that allow recipes.

afterthought2 committed:
Merge pull request #2354 from MistakeNot4892/recipes

Shuffling around the hardness values that allow recipes.

MistakeNot4892 committed:
Lowers the threshold for cleaners to act on turfs.

MistakeNot4892 committed:
Cleared access from tradeship and generic suit cyclers fully.

MistakeNot4892 committed:
Gated around a runtime for non-machine interfaces.

MistakeNot4892 committed:
Merge pull request #2330 from PsyCommando/Fix-heartbat-on-rejuv

Rejuvenating the heart should now fix it stopping

NebulaSS13Bot committed:
Automatic changelog generation for PR #2330 [ci skip]

MistakeNot4892 committed:
Linter pls.

afterthought2 committed:
Merge pull request #2324 from MistakeNot4892/codex

Codex ID overlap unit test.

MistakeNot4892 committed:
Adding safety checking to fluid turf and overlay queuing.

afterthought2 committed:
Merge pull request #2350 from MistakeNot4892/vendor

Removing broken vendor icon.

NebulaSS13Bot committed:
Automatic changelog generation [ci skip]

NebulaSS13Bot committed:
Automatic changelog generation for PR #2262 [ci skip]

MistakeNot4892 committed:
Tweaking GC changes.

MistakeNot4892 committed:
Lazylisting bad_external_organs while we're in the neighborhood.

MistakeNot4892 committed:
Reapplying a dangling ref fix to credits from Hearth.

MistakeNot4892 committed:
Reapplying a GC from hearth again.

NebulaSS13Bot committed:
Automatic changelog generation for PR #2352 [ci skip]

quardbreak committed:
Fixes space in singular name variable

afterthought2 committed:
Simplifies and modernizes report access.

afterthought2 committed:
Merge pull request #2359 from MistakeNot4892/hearthfixes

More GC fixes and tweaks from Hearth.

afterthought2 committed:
Merge pull request #2360 from MistakeNot4892/hearthfixes2

Lazylists bad_external_organs, applies ref fixes from Hearth.

NebulaSS13Bot committed:
Automatic changelog generation [ci skip]

noelle-lavenza committed:
Fix GC issues with temperature (+ maybe reactions)

quardbreak committed:
Replaces simply dmi preview with AnturK's dmi editor vscode extension

MistakeNot4892 committed:
Swaps direct level list references out for procs.

MistakeNot4892 committed:
Merge pull request #2363 from afterthought2/report_access_modernization

Simplifies and modernizes report access.

afterthought2 committed:
Merge pull request #2367 from quardbreak/anturk_dmieditor

Replaces simply dmi preview with AnturK's dmi editor vscode extension

afterthought2 committed:
Merge pull request #2365 from jade-lavenza/nebula-fix/gc-startup

Fix GC issues with temperature (+ maybe reactions)

afterthought2 committed:
Additional debug information included in power tally unit test.

afterthought2 committed:
Reduces thruster power usage.

afterthought2 committed:
More intuitive airlock/blast door construction

quardbreak committed:
Removes outdated comment

PsyCommando committed:
Attempt at applying requested changes

* Added proc to give feedback to user on why they cannot place a cable.
* Changed the logic a bit.

MistakeNot4892 committed:
Merge pull request #2370 from quardbreak/no_comments

Removes outdated comment

NebulaSS13Bot committed:
Automatic changelog generation for PR #2368 [ci skip]

MistakeNot4892 committed:
Implementing suggested reordering of AM Destroy()

NebulaSS13Bot committed:
Automatic changelog generation [ci skip]

PsyCommando committed:
Limb dislocation fix + improvements (#2337)

* Replaced checks for dislocation

A lot checks for dislocation were comparing against 2 for some reasons, which would never be true.
Replaced all the other checks with a call to is_dislocated()

* Replaced all dislocated check with is_dislocated

* Dislocation immunity set by limb flag now

Dislocation used to use magic numbers to determine if a limb was dislocated, not dislocated or immune to dislocation. While everything else used a limb flag to define this kind of thing. So I made a flag for dislocation instead to bring this stuff closer to the rest of the organ "capabilities"

* un-dislocate limbs on rejuv

* fix typo

* Added pain to dislocate

For some reasons, none of the code would actually cause any sort of pain when dislocating a limb.. Except when un-dislocating. So added some pain + a little bit of weaken, since that hurts a whole lot.

* Added feedback on skillcheck failure

When dislocating a limb, if the grabber doesn't have the required skills it would just say nothing and fail. Even though the current code always warn you you don't have a strong enough grip to dislocate regardless of your skills.
So I added a warning when failing to joint lock or dislocate.

* Change weaken to stun

* Didn't mean to commit this line

* Applied suggested text fixes

* Made dislocation a status flag

* Fix the oopsie I made when rebasing

* Replaced checks for dislocation

A lot checks for dislocation were comparing against 2 for some reasons, which would never be true.
Replaced all the other checks with a call to is_dislocated()

* Replaced all dislocated check with is_dislocated

* un-dislocate limbs on rejuv

* Fix the fix that was fixed before my fix

* The rebase apparently re-duplicated this for some reasons

* Use limb level can_feel_pain proc

NebulaSS13Bot committed:
Automatic changelog generation for PR #2337 [ci skip]

NebulaSS13Bot committed:
Automatic changelog generation for PR #2342 [ci skip]

afterthought2 committed:
Changes dmdoc workflow to forbid force-pushing.

MistakeNot4892 committed:
Merge pull request #2371 from afterthought2/stop_the_spam

Changes dmdoc workflow to forbid force-pushing.

NebulaSS13Bot committed:
Automatic changelog generation [ci skip]

MistakeNot4892 committed:
Adding a roleplay summary var to species preview.

PsyCommando committed:
fix cable messages grammar

MistakeNot4892 committed:
Converting tool macros to use the correct casing.

afterthought2 committed:
Merge pull request #2333 from PsyCommando/Add-feedback-when-wiring-zcable

Add feedback when wiring zcable + tweaks

NebulaSS13Bot committed:
Automatic changelog generation for PR #2333 [ci skip]

NebulaSS13Bot committed:
Automatic changelog generation [ci skip]

NebulaSS13Bot committed:
Automatic changelog generation [ci skip]

MistakeNot4892 committed:
Splitting serpentids into their own modpack.

quardbreak committed:
Removes AS_ANYTHING macro due no need

MistakeNot4892 committed:
Flash refactor.

kelgaras committed:
new sprites for vents and scrubbers

afterthought2 committed:
Merge pull request #2397 from quardbreak/as_anything_outdated

Removes AS_ANYTHING macro due no need

afterthought2 committed:
Merge pull request #2372 from keIgaras/vents-ans-scrubs

new sprites for vents and scrubbers

NebulaSS13Bot committed:
Automatic changelog generation for PR #2372 [ci skip]

MistakeNot4892 committed:
Swaps damtype strings out for defines.

MistakeNot4892 committed:
Implemented nopain todo.

MistakeNot4892 committed:
Implements distance check todo in storage.

NebulaSS13Bot committed:
Automatic changelog generation [ci skip]

afterthought2 committed:
Merge pull request #2366 from MistakeNot4892/maplists

Swaps direct level list references out for procs.

afterthought2 committed:
Merge pull request #2396 from MistakeNot4892/snek

Splitting serpentids into their own modpack.

afterthought2 committed:
Merge pull request #2398 from MistakeNot4892/flashes

Flash refactor.

afterthought2 committed:
Merge pull request #2403 from MistakeNot4892/todo2

Implements two small todo comments.

afterthought2 committed:
Merge pull request #2404 from MistakeNot4892/todo3

Swaps damtype strings out for defines.

afterthought2 committed:
Merge pull request #2208 from MarinaGryphon/bugfix/lowercase-names

Convert a lot of improper noun names to lowercase

MistakeNot4892 committed:
Converted coloured square HTML to macro.

MistakeNot4892 committed:
Changed a handful of direct filter setting to use proc.

MistakeNot4892 committed:
Added equip hook for borgs.

MistakeNot4892 committed:
Added z-level name retrieval proc.

MistakeNot4892 committed:
Cleaned up overlooked tags.

Commit:
fixfixfix

PsyCommando committed:
Initial

Moved saved var to within the dm files, since its tightly coupled to the actual code, and it allows the compiler to verify vars at compile time and make my life less of a living hell.

PsyCommando committed:
Initial

Moved saved var to within the dm files, since its tightly coupled to the actual code, and it allows the compiler to verify vars at compile time and make my life less of a living hell.

PsyCommando committed:
Make components less silly

PsyCommando committed:
Make components less silly

afterthought2 committed:
Merge pull request #2383 from MistakeNot4892/toolcasing

Converting tool macros to use the correct casing.

afterthought2 committed:
Merge pull request #2335 from MistakeNot4892/inflatables

Minor inflatable refactor.

NebulaSS13Bot committed:
Automatic changelog generation for PR #2335 [ci skip]

MistakeNot4892 committed:
Merge pull request #2413 from SolatK/dev

boring way to fix

NebulaSS13Bot committed:
Automatic changelog generation [ci skip]

Commit:
fix: add mover check to `AdjustMovementDirection` proc. Add `mover` argument

NebulaSS13Bot committed:
Automatic changelog generation [ci skip]

afterthought2 committed:
Merge pull request #2424 from Gaxeer/fix-pulling-stunned-mob

fix: adjust mob's move direction, only if it is an initiator of movement

afterthought2 committed:
Merge pull request #2411 from MistakeNot4892/square

Small changes from #2204

PsyCommando committed:
Separated a bunch of things to files

Some of the saved variables related files were getting very crowded and messy. So moved things around a bit to place that made more sense.

PsyCommando committed:
Separated a bunch of things to files

Some of the saved variables related files were getting very crowded and messy. So moved things around a bit to place that made more sense.

PsyCommando committed:
Remove useless saved vars

* Removed vars that are always constant or setup by code.
* Removed vars that are temporary in nature.
* Removed vars that can't be written to. (X,Y,Z)

PsyCommando committed:
Remove useless saved vars

* Removed vars that are always constant or setup by code.
* Removed vars that are temporary in nature.
* Removed vars that can't be written to. (X,Y,Z)

PsyCommando committed:
Forgot to push saved vars optimization macro

PsyCommando committed:
Forgot to push saved vars optimization macro

PsyCommando committed:
Made decals saveable for now

* Since some things are decals that shouldn't go away made them saved. Probably should be moved to a better subclass one day.

PsyCommando committed:
Made decals saveable for now

* Since some things are decals that shouldn't go away made them saved. Probably should be moved to a better subclass one day.

PsyCommando committed:
Added a type check to the resolver

After getting bad cast errors I added a check to skip over things that aren't the expected type until we can work this out.

PsyCommando committed:
Added a type check to the resolver

After getting bad cast errors I added a check to skip over things that aren't the expected type until we can work this out.

PsyCommando committed:
Used the recommended way of adding overlays

After loading from save, turf decals were added straight up to the overlays. Used add_overlay instead since it does a bit more in a less risky way.

PsyCommando committed:
Used the recommended way of adding overlays

After loading from save, turf decals were added straight up to the overlays. Used add_overlay instead since it does a bit more in a less risky way.

PsyCommando committed:
Removed pointless saved decals stuff

* Overlays is cleared on update_icon, so this was just adding it, before it got immediately deleted. update_icon already does it for us anyways.

PsyCommando committed:
Removed pointless saved decals stuff

* Overlays is cleared on update_icon, so this was just adding it, before it got immediately deleted. update_icon already does it for us anyways.

PsyCommando committed:
Fixed a few annoyances with the kleibkhar map

* Changed all wallframes to be the default material, since paint doesn't work on titanium.
* Added missing hatch decals under some doors.
* Moved signs in the cargo docking area onto the adjacent interior turf so that they're not visible from outside the enclosure.

PsyCommando committed:
Removed unused saved var stuff

PsyCommando committed:
Removed unused saved var stuff

PsyCommando committed:
Updated text2path uses in tests

PsyCommando committed:
Updated text2path uses in tests

PsyCommando commented:
Just to clear things up. This is mostly ready for review. I just need to see if the longer saving time issue is introduced by this, or if its a by product of the map changes.

PsyCommando committed:
Removed saved vars for the species mod

Since some maps don't include those species it runtimes in the unit tests.
I didn't move them because they moved the species again upstream and it would cause a conflict, so removed for now.

PsyCommando committed:
Removed saved vars for the species mod

Since some maps don't include those species it runtimes in the unit tests.
I didn't move them because they moved the species again upstream and it would cause a conflict, so removed for now.

PsyCommando committed:
Removed useless saved vars for the solar control

All this stuff is reset on init and doesn't need to be saved at all.

PsyCommando committed:
Removed useless saved vars for the solar control

All this stuff is reset on init and doesn't need to be saved at all.

PsyCommando committed:
Added instance count to the map saving statistics

PsyCommando committed:
Added instance count to the map saving statistics

PsyCommando committed:
forgot to commit these lines

PsyCommando committed:
forgot to commit these lines

PsyCommando committed:
Fixed multi-tile objects saving

PsyCommando committed:
Fixed multi-tile objects saving

PsyCommando committed:
Fixed solars duplicating materials on init

PsyCommando committed:
Fixed solars duplicating materials on init

PsyCommando committed:
Fix doors being stupid

PsyCommando committed:
Fix doors being stupid

NataKilar committed:
Merge pull request #301 from PsyCommando/minor_map_fix

Fixed a few annoyances with the kleibkhar map

NataKilar committed:
Merge remote-tracking branch 'upstream/dev' into may-2022-merge

NataKilar committed:
May 2022 merge fixes

PsyCommando committed:
Skip saving some useless stuff

PsyCommando committed:
Skip saving some useless stuff

PsyCommando committed:
Removed more useless saved vars

* Also removed all things flattened since the flattener is broken and causing weird things to happen to lists..

PsyCommando committed:
Removed more useless saved vars

* Also removed all things flattened since the flattener is broken and causing weird things to happen to lists..

PsyCommando committed:
Added better diagnostic output on save

* Errors when saving lists are a lot more verbose and useful.
* Now reports the time it takes in total to do saved vars lookup.

PsyCommando committed:
Added better diagnostic output on save

* Errors when saving lists are a lot more verbose and useful.
* Now reports the time it takes in total to do saved vars lookup.

PsyCommando committed:
Forgot to commit this

PsyCommando committed:
Forgot to commit this

PsyCommando committed:
Sorted time spent per type list

PsyCommando committed:
Sorted time spent per type list

PsyCommando committed:
Fixed bad area for kleibkhar

PsyCommando committed:
Fixed bad area for kleibkhar

PsyCommando committed:
Set base turf for all kleibkhar areas.

* Added is_outside to outside areas, and a few flags too

PsyCommando committed:
Set base turf for all kleibkhar areas.

* Added is_outside to outside areas, and a few flags too

PsyCommando committed:
Fixed kleibhkar temp so its static now

PsyCommando committed:
Fixed kleibhkar temp so its static now

PsyCommando committed:
remove useless code

PsyCommando committed:
remove useless code

PsyCommando committed:
runtimes fixes

PsyCommando committed:
runtimes fixes

PsyCommando committed:
Changed area on kleibkhar 4 to be a "sky" area

Since base turfs can only be overriden via areas had to do this to prevent ground turfs being created in the sky.

PsyCommando committed:
Changed area on kleibkhar 4 to be a "sky" area

Since base turfs can only be overriden via areas had to do this to prevent ground turfs being created in the sky.

PsyCommando committed:
Skip saving areas marked as non persistent

PsyCommando committed:
Skip saving areas marked as non persistent

PsyCommando committed:
Fix turfs on exoplanets loosing their areas

Since we're not saving all turf non-exterior turfs on planets lose their areas unless we do this..

PsyCommando committed:
Fix turfs on exoplanets loosing their areas

Since we're not saving all turf non-exterior turfs on planets lose their areas unless we do this..

NataKilar commented on mods/persistence/controllers/subsystems/persistence.dm:
```diff
@@ -199,9 +202,20 @@
+					//Ignore non-saved areas
+					var/area/TA = T.loc 
+					if(istype(TA) && (TA.area_flags & AREA_FLAG_IS_NOT_PERSISTENT))
+						for(var/mob/living/L in T)
```
This should be added to the check in the loop which adds player minds to limbo if they're in an unsaved area for consistency, rather than being here.

PsyCommando replied:
Speaking of that, does your mind stuff actually works? Because last time I checked, if you save while you're in a non saved area, you just cannot spawn on the menu. I told brawler about it, but I don't think he knows how to fix that.
But yeah, I can just remove that from there.

NataKilar commented on mods/persistence/game/machinery/machinery.dm:
```diff
@@ -11,4 +11,73 @@
+SAVED_VAR(/obj/machinery, uid)
+SAVED_VAR(/obj/machinery, panel_open)
+SAVED_VAR(/obj/machinery, uncreated_component_parts)
+//SAVED_VAR(/obj/machinery, component_parts)
```
Maybe add a comment here about the explicit exclusion.

PsyCommando replied:
It probably just shouldn't be in there tbh. Machinery fixing stuff is on another branch currently. Not sure why I didn't remove that.

NataKilar commented on mods/persistence/__defines/saved_var.dm:
```diff
@@ -0,0 +1,21 @@
+#define SAVED_VAR(CLASSPATH, VARNAME) CLASSPATH/proc/zz_compilerVarExistsTest_##VARNAME(){.=src.##VARNAME;}; /decl/saved_variables##CLASSPATH/make_saved_variables(list/saved){LAZYDISTINCTADD(saved, #VARNAME); return ..(saved);}
+
+/**
+ * Clears all saved variables for this particular type and its children.
```
Might want to mention that it only saves its position/existence or whatever. Makes its purpose more explicit.

PsyCommando replied:
I guess I wrote it at the wrong place. I'll add a copy in there.

NataKilar commented on mods/persistence/controllers/subsystems/persistence.dm:
```diff
@@ -1,3 +1,6 @@
+/proc/cmp_serialization_stats_dsc(var/datum/serialization_stat/S1, var/datum/serialization_stat/S2)
```
Nit, but I'd prefer this to be at the bottom of the file

NataKilar commented on mods/persistence/datums/wires/wires.dm:
```diff
@@ -0,0 +1,5 @@
+//Fix annoyance
```
Comment is unclear, probably not necessary.

PsyCommando replied:
The comment is explained on line 3

NataKilar commented on mods/persistence/game/machinery/machinery.dm:
```diff
@@ -11,4 +11,73 @@
+
+/obj/machinery/power/tracker/after_deserialize()
+	. = ..()
+	connect_to_network() //There's a powernet mismatch sometimes after loading the tracker
```
Does this do anything? Trackers already call connect_to_network in Initialize(), and if the powernets somehow change between this and Initialize(), it's going to be a member of two seperate powernets.

PsyCommando replied:
I don't think that was meant to be commited actually.. Might have to revert that. 
Solars are way more busted than that could ever fix, and I've been working on something bigger. Its busted on the upstream level at that.

NataKilar commented on mods/persistence/game/objects/items/items.dm:
```diff
@@ -0,0 +1,16 @@
+SAVED_VAR(/obj/item, health)
+
+//Used by things that can be collapsed and put into other inv slots
+SAVED_VAR(/obj/item, slot_flags)
```
It can wait, but this should really be set on specific subtypes that have this behavior. Might want to add a TODO

PsyCommando replied:
Eh, I think people can figure it out. If we add todo everywhere people will even care less about them.

NataKilar commented on mods/persistence/game/objects/items/items.dm:
```diff
@@ -0,0 +1,16 @@
+
+//Used by clothing that covers or not the face and can be toggled
+SAVED_VAR(/obj/item, flags_inv)
+SAVED_VAR(/obj/item, body_parts_covered)
```
Ditto

NataKilar commented on mods/persistence/modules/world_save/_persistence.dm:
```diff
@@ -10,88 +9,36 @@ var/global/list/custom_saved_lists = list() // Custom saved lists are kept here
-		coord_list.Add("[T.x],[T.y],[T.z]")
-	return coord_list
+/**Returns a list of saved variables for this entity type. */
+/datum/proc/get_saved_vars()
```
Is this called anywhere? Seems extraneous with ``get_saved_variables_for()`` but if you think it's useful it can be kept.

PsyCommando replied:
Its possible I forgot to remove it. Its mainly a leftover from the old system, which actually didn't even need it since vars were stored globally then too.

NataKilar commented on mods/persistence/modules/world_save/saved_vars/saved_skipped.dm:
```diff
@@ -0,0 +1,33 @@
+/mob/living/simple_animal/hostile/mimic
+	should_save = FALSE
+
+/datum/extension/penetration
```
The extension overrides are unnecessary, should_save is set to FALSE by default for extensions.

PsyCommando replied:
I've had a problem with that. They would save and cause problems.

NataKilar replied:
That doesn't really make sense, should_save on the base type ``/datum/extension`` is FALSE, unless it's set elsewhere for some bizarre reason. At a glance though, it shouldn't be.

NataKilar commented on mods/persistence/modules/world_save/serializers/_serializer.dm:
```diff
@@ -56,6 +56,9 @@
 	// We check to see if this is a reference only var so that if things are missing from the resolver, this doesn't fail silently.
 	if(reference_only && !resolver.things["[object_id]"])
 		return null
+	if(!istype(resolver.things["[object_id]"], /datum/persistence/load_cache/thing))
```
When, if ever, would this occur?

PsyCommando replied:
Every times I'd load from save actually? 
Kept getting things trying to access non-existent vars there. And I couldn't explain it. Its like there's shared ids.

NataKilar commented on mods/persistence/modules/world_save/serializers/sql_serializer.dm:
```diff
@@ -146,7 +159,7 @@
 
-	for(var/V in object.get_saved_vars())
+	for(var/V in get_saved_variables_for(object.type))
 		if(!(V in object.vars))
```
As per our discussion on discord, this needs to be removed so that the save doesn't slow to a crawl

PsyCommando replied:
Alright I guess I'll make the change myself then.

NataKilar committed:
Saving time fix

PsyCommando committed:
Merge pull request #304 from NataKilar/may-2022-merge

May 2022 merge [IDB IGNORE] [MDB IGNORE]

NataKilar committed:
Fixes limbo saving in non-saved areas

PsyCommando committed:
Merge pull request #305 from NataKilar/saving-time-fix

Saving time fix

PsyCommando committed:
fix missing area

Deleted this area by accident. Its only purpose is to fool the pretty lame unit tests enforces contents of areas.

PsyCommando committed:
fix missing area

Deleted this area by accident. Its only purpose is to fool the pretty lame unit tests enforces contents of areas.

PsyCommando committed:
Applied changes

PsyCommando committed:
Applied changes

PsyCommando committed:
Switch back to creating turfs from scratch

Using change turf causes artifacts and more crap to happen.

PsyCommando committed:
Switch back to creating turfs from scratch

Using change turf causes artifacts and more crap to happen.

PsyCommando committed:
fixed bad saved vars

May need to update those later!

PsyCommando committed:
Merge branch 'saved_var_refactor' of https://github.com/PsyCommando/Nebula into saved_var_refactor

PsyCommando committed:
Fixed file permission on check-paths.sh!

PsyCommando committed:
Increased text2path uses to 38 since the merge increased it..

NataKilar commented on mods/persistence/controllers/subsystems/persistence.dm:
```diff
@@ -101,7 +101,11 @@
-			if(!QDELETED(current_mob) && ((current_mob.z in SSpersistence.saved_levels) || (get_area(current_mob) in SSpersistence.saved_areas)))
+			var/area/MA = get_area(current_mob)
+			if(!QDELETED(current_mob) && istype(MA) && \
+				(!(MA.area_flags & AREA_FLAG_IS_NOT_PERSISTENT) || \
```
This check isn't correct. It needs to be ``if(!(MA.area_flags & AREA_FLAG_IS_NOT_PERSISTENT) && (MA in SSpersistence.saved_areas || current_mob.z in SSpersistence.saved_levels))``

NataKilar commented on mods/persistence/modules/world_save/serializers/sql_serializer.dm:
```diff
@@ -408,52 +399,53 @@ var/global/list/serialization_time_spent_type
+
+	//Remove vars not in existing.vars we got in thing.thing_vars
+	for(var/datum/persistence/load_cache/thing_var/TV in thing.thing_vars)
+		if(!(TV.key in existing.vars))
```
I know you just moved this from the place below, but instead of doing this, try doing
```
try
    var/V = existing.vars[TV.key]
catch
    log_warning(blah)
```
It should be much faster.

PsyCommando replied:
That's really not faster? Exceptions are the most expensive thing to handle. Its fast because you probably didn't try it with missing variables, but once it starts throwing exceptions its gonna take centuries to load. Don't ever do flow control with exceptions.

PsyCommando replied:
I found a way better way to handle them. Just check the loaded vars against the saved vars. Since they're checked at compile time they can't refer to non-existent vars, and its just a regular, fast list lookup.

PsyCommando committed:
implemented fixes
